server/webapp/WEB-INF/rails.new/app/controllers/api_v1/admin/security/roles_controller.rb

```diff
@@ -0,0 +1,104 @@
+
+module ApiV1
+  module Admin
+    module Security
+      class RolesController < ::ApiV1::BaseController
+        before_filter :check_admin_user_and_401
+        before_action :check_for_stale_request, :check_for_attempted_rename, only: [:update]
+
+        def index
+          roles = service.listAll().to_a
+
+          if params[:role_type] == 'role'
+            roles.keep_if { |role| role.kind_of?(RoleConfig) }
+          elsif params[:role_type] == 'plugin_role'
+            roles.keep_if { |role| role.kind_of?(PluginRoleConfig) }
+          elsif params[:role_type].present?
+            return render_message("Bad role type `#{params[:role_type]}`. Valid values are `role` and `plugin_role`", :bad_request)
+          end
+
+          render DEFAULT_FORMAT => ApiV1::Security::RolesConfigRepresenter.new(roles.to_a).to_hash(url_builder: self)
+        end
+
+        def show
+          role = load_entity_from_config
+          if stale?(etag: etag_for(role))
+            render DEFAULT_FORMAT => ApiV1::Security::RolesConfigRepresenter.representer_instance_for(role).to_hash(url_builder: self)
+          end
+        end
+
+        def update
+          role = load_entity_from_config
+          role_from_request = ApiV1::Security::RolesConfigRepresenter.representer_instance_from_hash(params[:role])
+
+          result = HttpLocalizedOperationResult.new
+          service.update(current_user, etag_for(role), role_from_request, result)
+          handle_create_or_update_response(result, role_from_request)
+        end
+
+        def create
+          result = HttpLocalizedOperationResult.new
+          role = ApiV1::Security::RolesConfigRepresenter.representer_instance_from_hash(params[:role])
+          service.create(current_user, role, result)
+          handle_create_or_update_response(result, role)
+        end
+
+        def destroy
+          result = HttpLocalizedOperationResult.new
+          service.delete(current_user, load_entity_from_config, result)
+          render_http_operation_result(result)
+        end
+
+        protected
+
+        def service
+          role_config_service
+        end
+
+        def load_entity_from_config
+          role_config_service.findRole(params[:role_name]) || (raise ApiV1::RecordNotFound)
+        end
+
+        def stale_message
+          LocalizedMessage::string('STALE_RESOURCE_CONFIG', 'role', params[:role_name])
+        end
+
+        def etag_for_entity_in_config
+          etag_for(load_entity_from_config)
+        end
+
+        def handle_create_or_update_response(result, updated_role)
+          json = ApiV1::Security::RolesConfigRepresenter.representer_instance_for(updated_role).to_hash(url_builder: self)
+          if result.isSuccessful
+            response.etag = [etag_for(updated_role)]
+            render DEFAULT_FORMAT => json
+          else
+            render_http_operation_result(result, {data: json})
+          end
+        end
+
+        def check_for_attempted_rename
```

Comment:
Do we need to `check_for_attempted_type_change` to avoid changing type of roles? I guess probably not, but just putting it out there.